### PR TITLE
Query surface formats

### PR DESCRIPTION
**Pull request by msiglreith**
Add functionality to query supported surface formats for swapchain creation.

- [x] hal
- [x] vulkan
- [x] dx12
- [x] metal
- [x] gl
- [x] render
- [x] hal/quad
- [x] render/example

### CODE REVIEW

File: src/backend/vulkan/src/conv.rs [hunk 1]
```diff
@@ -148,6 +148,203 @@ pub fn map_format(surface: format::SurfaceType, chan: format::ChannelType) -> Op
     })
 }
 
+pub fn map_vk_format(
+    fmt: vk::Format
+) -> Option<(format::SurfaceType, format::ChannelType)> {
+    use ash::vk::Format::*;
+    use core::format::SurfaceType::*;
+    use core::format::ChannelType::*;
+    match fmt {
```
Comment (kvark) [inline src/backend/vulkan/src/conv.rs]: it's easier to write `Some(match fmt {` here, and then only `return None` in some of the branches

File: src/backend/vulkan/src/window.rs [hunk 2]
```diff
@@ -7,6 +7,7 @@ use ash::vk;
 use ash::extensions as ext;
 
 use hal;
+use core::format::Format;
```
Comment (kvark) [inline src/backend/vulkan/src/window.rs]: should be `hal`

File: src/hal/src/window.rs [hunk 3]
```diff
@@ -218,7 +221,20 @@ impl SwapchainConfig {
...
+
+
+    /// Specify the color format for the backbuffer images.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    ///
+    /// ```
+    pub fn with_color_typed<Cf: format::RenderFormat>(mut self) -> Self {
```
Comment (kvark) [inline src/hal/src/window.rs]: nice, looks like we are establishing a convention
Now I wonder if mapping API should follow (instead of `_raw`)

**Comment by msiglreith** [comment]
Thanks! Merged `capabilities` and `swapchain_formats` into `capabilities_and_formats`

**Comment by msiglreith** [comment]
bors r=kvark

**Pull request merged**